test/jdk/java/net/httpclient/ExpectContinueTest.java

```diff
@@ -0,0 +1,296 @@
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Tests that when the httpclient sends a 100 Expect Continue header and receives
+ *          a response code of 417 Expectation Failed, that the client does not hang
+ *          indefinitely and closes the connection.
+ * @bug 8286171
+ * @modules java.base/sun.net.www.http
+ *          java.net.http/jdk.internal.net.http.common
+ *          java.net.http/jdk.internal.net.http.frame
+ *          java.net.http/jdk.internal.net.http.hpack
+ *          jdk.httpserver
+ * @library /test/lib http2/server
+ * @compile HttpServerAdapters.java
+ * @run testng/othervm ExpectContinueTest
+ */
+
+
+import com.sun.net.httpserver.HttpServer;
+
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import javax.net.ServerSocketFactory;
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.io.Writer;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.StringTokenizer;
+import java.util.concurrent.CompletableFuture;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.testng.Assert.assertEquals;
+
+public class ExpectContinueTest implements HttpServerAdapters {
+
+    HttpTestServer http1TestServer; // HTTP/1.1
+    Http1HangServer http1HangServer;
+    HttpTestServer http2TestServer; // HTTP/2
+
+    URI getUri;
+    URI postUri;
+    URI hangUri;
+    URI h2getUri;
+    URI h2postUri;
+    URI h2hangUri;
+
+    static final String EXPECTATION_FAILED_417 = "417 Expectation Failed";
+
+    @BeforeTest
+    public void setup() throws Exception {
+        InetSocketAddress sa = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+        InetSocketAddress saHang = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+
+        http1TestServer = HttpTestServer.of(HttpServer.create(sa, 0));
+        http1TestServer.addHandler(new GetHandler(), "/http1/get");
+        http1TestServer.addHandler(new PostHandler(), "/http1/post");
+        getUri = URI.create("http://" + http1TestServer.serverAuthority() + "/http1/get");
+        postUri = URI.create("http://" + http1TestServer.serverAuthority() + "/http1/post");
+
+        // Due to limitations of the above Http1 Server, a manual approach is taken to test the hanging with the
+        // httpclient using Http1 so that the correct response header can be returned for the test case
+        http1HangServer = new Http1HangServer(saHang);
```

Comment:
It took me a while to understand why we have special server implementation for this. The comment helped :) It reminded me about the `ServerImpl` used by the `HttpServer`, which unconditionally sends a `100` response code when it sees a `Expect: 100-continue` in the request header.

Reply:
Yep, thats spot on. That is the exact case which made it necessary to include our own Server implementation for the HTTP/1 417 scenario.